Add the ability to compile with PGO

nathanchance opened the pull request:
[Profile guided optimization](https://llvm.org/docs/HowToBuildWithPGO.html) can help a lot with performance,
especially when the use case for the generated toolchain is very
specific like ours (we only care about compiling the kernel).

This adds a flag that users can enable, which will build a bootstrap
compiler (stage 1), an instrumented compiler (stage 2), a set of kernels
with the instrumented compiler, and then use the generate profiles to
build the final instrumented compiler.

To test the difference with and without PGO, I built a regular LLVM
toolchain:

`$ ./build-llvm.py`

and a PGO enabled one (using the same revision):

`$ ./build-llvm.py --pgo --no-pull`

and built the following kernels:

* arm32 allyesconfig
* arm64 allyesconfig
* powerpc powernv_defconfig
* powerpc ppc44x_defconfig
* x86_64 allyesconfig

The compile times for both the compilers and the kernels are below:

Clang build (without PGO): 14m 17s
Clang build (with PGO): 39m 41s

Kernel build (without PGO): 40m 51s
Kernel build (with PGO): 33m 13s

While the clang compilation time significantly increases, there is
almost a 20% reduction in kernel build time, which can add up a lot in
daily development (plus it is hidden behind an option by default).

MSF-Jarvis commented on build-llvm.py:
```diff
@@ -315,6 +322,28 @@ def cleanup(build_folder, incremental):
+    :param args: The args variable generated by parse_parameters
+    :return: The final stage number
+    """
+    if args.pgo:
```
```suggestion
    3 if args.pgo else 2
```

Just a personal preference

nathanchance replied:
Nice! Done locally, will push after addressing all comments.

MSF-Jarvis commented on build-llvm.py:
```diff
@@ -315,6 +322,28 @@ def cleanup(build_folder, incremental):
+def instrumented_stage(args, stage):
+    """
+    Returns true if we are using PGO and on stage 2
+    :param args: The args variable generated by parse_parameters
```
Maybe just pass `args.pgo` as a boolean?

nathanchance replied:
I think it is easier to have the `args.pgo` in here, rather than in all of the call spots but I don't feel strongly about it.

MSF-Jarvis replied:
Yeah I don't either, let's just keep `args`.

MSF-Jarvis commented on build-llvm.py:
```diff
@@ -385,32 +414,42 @@ def cc_ld_cmake_defines(dirs, env_vars, stage):
         defines['LLVM_USE_LINKER'] = ld
 
-    # Use llvm-ranlib for stage 2 builds
+    if llvm_tblgen is not None:
```
```suggestion
    if llvm_tblgen:
```

MSF-Jarvis replied:
```
Python 3.6.8 (default, Jan 14 2019, 11:02:34) 
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> llvm_tblgen = None
>>> if llvm_tblgen:
...     print("llvm_tblgen is not None")
... else:
...     print("llvm_tblgen is None")
... 
llvm_tblgen is None
```

nathanchance replied:
Will do that for all of the if statements.

MSF-Jarvis commented on build-llvm.py:
```diff
@@ -385,32 +414,42 @@ def cc_ld_cmake_defines(dirs, env_vars, stage):
+        defines['LLVM_TABLEGEN'] = llvm_tblgen
+
+    # Use llvm-ranlib for stage 2+ builds
     if ranlib is not None:
```
```suggestion
    if ranlib:
```

MSF-Jarvis commented on kernel/build.sh:
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env bash
```
Managed to sneak in bash after all! :stuck_out_tongue_winking_eye:

nathanchance commented:
Don't merge this just yet, @MSF-Jarvis  pointed out that the `kernel/build.sh` assumes that the user is not changing the default targets of `build-llvm.py`; if they do, `kernel/build.sh` will fail because some of the builds will fail due to an unsupported backend. The script will need to accept the targets as a parameter.

nickdesaulniers commented:
@pirama-arumuga-nainar is our resident expert on PGO

pirama-arumuga-nainar commented on build-llvm.py:
```diff
@@ -576,13 +621,36 @@ def invoke_ninja(args, dirs, stage):
+    # Combine profiles
+    stage2_folder = dirs.build_folder.joinpath("stage2")
+    subprocess.run([
+        stage2_folder.joinpath("bin", "llvm-profdata"), "merge",
```
Using the stage2 llvm-profdata may slow this step down since that is also built with instrumentation.  You can safely use the llvm-profdata from stage1.

nathanchance replied:
Thanks for the review, I’ll make that change tonight.

stephenhines commented:
> Clang build (without PGO): 14m 17s
> Clang build (with PGO): 39m 41s

This is quite shocking, and I am quite surprised that it would behave this badly. Considering that you do bootstrap Clang, is it worth putting Clang in the PGO profiling set? The only way I could see this being this slow is if literally everything related to C++ (not C) shows up as unused/cold because the kernel is C-only. Still, seeing a 2x slowdown is surprising for that case.

stephenhines commented:
> > Clang build (without PGO): 14m 17s
> > Clang build (with PGO): 39m 41s
> 
> This is quite shocking, and I am quite surprised that it would behave this badly. Considering that you do bootstrap Clang, is it worth putting Clang in the PGO profiling set? The only way I could see this being this slow is if literally everything related to C++ (not C) shows up as unused/cold because the kernel is C-only. Still, seeing a 2x slowdown is surprising for that case.

Pirama pointed out that I misread this as you compiling Clang itself slower, when it is really this build script that is doing more work now. I think that is expected, and probably worthwhile if you are building multiple kernels and/or reusing your Clang builds for multiple kernel builds.

nathanchance commented:
> > Clang build (without PGO): 14m 17s
> > Clang build (with PGO): 39m 41s
> 
> This is quite shocking, and I am quite surprised that it would behave this badly. Considering that you do bootstrap Clang, is it worth putting Clang in the PGO profiling set? The only way I could see this being this slow is if literally everything related to C++ (not C) shows up as unused/cold because the kernel is C-only. Still, seeing a 2x slowdown is surprising for that case.

I should have probably been clearer about that. That time includes the three stages + the profiling (which takes about 15-20 minutes on my machine because it is five kernel builds).

nathanchance commented:
Alright, I added Pirama's suggestion, fixed the issue that Harsh brought up and reworded the commit message to make it clear that `build-llvm.py` is what takes longer, not the individual clang compiles. I think this should be good to go, unless others have any comments, questions, or concerns :)

nathanchance commented:
Thanks for all of the review and testing, I really appreciate it.
